docs: add Grid Pro text and TS examples (#206)
vaadin/docs: articles/ds/components/grid-pro/index.asciidoc

---

= Grid Pro

> :experimental:

Author: It is needed due to the use of the keyboard macro.

> :experimental:

Reviewer: Yeah. We could set that attribute in docs-app already, but right now it’s needed in each article separately.

// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

.Features shared with Grid
[NOTE]
====

> Grid Pro is an extension of the <<../grid#,Grid>> component and thus all of Grid’s features

Reviewer: ```suggestion
Grid Pro is an extension of the <<../grid#,Grid>> component and all Grid’s features
```

----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProSingleClick.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

=== Single Cell Edit

By default, when in edit mode, kbd:[Tab] and kbd:[Shift+Tab] moves to the next and previous editable cell, respectively, while remaining in edit mode.

Single Cell Edit is a mode that makes tabbing from one cell to another exit edit mode.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/gridpro/grid-pro-single-cell-edit.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProSingleCellEdit.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

=== Enter Next Row

Pressing kbd:[Enter] and kbd:[Shift+Enter] saves the changes and exists edit mode by default.

> kbd:[Enter] and kbd:[Shift+Enter] can be made to additionally focus the editable cell in the next and previous row, respectively, by using the Enter Next Row mode.

Reviewer: ```suggestion
kbd:[Enter] and kbd:[Shift+Enter] can be made to focus the editable cell in the next and previous row, respectively, by using the Enter Next Row mode.
```

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/gridpro/grid-pro-enter-next-row.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProEnterNextRow.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

== Edit Column

Editing is enabled on a per column basis.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/gridpro/grid-pro-edit-column.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProEditColumn.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

=== Recommended Built-in Editors

> Grid Pro features three recommended built-in editors: Text Field, Checkbox and Select.

Reviewer: ```suggestion
Grid Pro features three recommended built-in editors: Text Field, Checkbox, and Select.
```

----
include::../../../../frontend/demo/component/gridpro/grid-pro-editors.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProEditor.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

== Prevent Saving Changes

You can rollback changes when the entered input is incorrect or invalid.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/gridpro/grid-pro-prevent-save.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProPreventSave.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

== Styling Editable Cells

Unless all columns are editable, apply styling to visually distinguish between editable and non-editable cells using the `editable-cell` part name selector.

> Additionally or alternatively, you can suffix read-only columns with “(read-only)” and/or editable columns with “(editable)”.

Reviewer: ```suggestion
For example, you can suffix read-only columns with “(read-only)” and/or editable columns with “(editable)”.
```

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/gridpro/grid-pro-styling-editable-cells.ts[render,tags=snippet,indent=0,group=TypeScript]
----

[source,css]
----
include::../../../../frontend/themes/docs/components/vaadin-grid-pro-editable-cells.css[tags=editable-style,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/gridpro/GridProStylingEditableCells.java[render,tags=snippet,indent=0,group=Java]
// ----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/domain/Person.java[group=Java,tags=snippet]
// ----

--

== Best Practises

> === Inline vs Non-inline Editing

Reviewer: ```suggestion
=== Inline vs Non-Inline Editing
```

Inline editing is recommended when:

- the user typically needs to make a lot of small changes to different items
- quick editing is important

Non-inline editing is preferable

- there’s a lot of columns/fields

> - users typically need to edit only one particular item at a time

Reviewer: ```suggestion
- users typically need to edit only one item at a time
```

- adding new items is common (as you might want to have edit and create modes work the same way, and creating new items with inline editing is not recommended with Grid Pro).

> - any of the editors need to be bigger than a simple field, such as a Text Area or multiselect field of any kind

Reviewer: ```suggestion
- any of the editors need to be bigger than a simple field, such as a Text Area or multi-select field of any kind
```
